packages/ember/tests/routing/decoupled_basic_test.js

```diff
@@ -326,4 +327,217 @@ class extends ApplicationTestCase {
       );
     });
   }
+
+  [`@test The route controller specified via controllerName is used in render`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+    });
+
+    this.add('route:home', Route.extend({
+      controllerName: 'myController',
+      renderTemplate() {
+        this.render('alternative_home');
+      }
+    }));
+
+    this.add('controller:myController', Controller.extend({
+      myValue: 'foo'
+    }));
+
+    this.addTemplate('alternative_home', '<p>alternative home: {{myValue}}</p>');
+
+
+    return this.visit('/').then(() => {
+      let homeRoute = this.applicationInstance.lookup('route:home');
+      let myController = this.applicationInstance.lookup('controller:myController');
+      let text = this.$('p').text();
+
+      assert.deepEqual(homeRoute.controller, myController,
+        'route controller is set by controllerName');
+
+      assert.equal(text, 'alternative home: foo',
+        'The homepage template was rendered with data from the custom controller');
+    });
+
+  }
+
+  [`@test The route controller specified via controllerName is used in render even when a controller with the routeName is available`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+    });
+
+    this.addTemplate('home', '<p>home: {{myValue}}</p>');
+
+    this.add('route:home', Route.extend({
+      controllerName: 'myController'
+    }));
+
+    this.add('controller:home', Controller.extend({
+      myValue: 'home'
+    }));
+
+    this.add('controller:myController', Controller.extend({
+      myValue: 'myController'
+    }));
+
+    return this.visit('/').then(() => {
+      let homeRoute = this.applicationInstance.lookup('route:home');
+      let myController = this.applicationInstance.lookup('controller:myController');
+      let text = this.$('p').text();
+
+      assert.deepEqual(homeRoute.controller, myController,
+        'route controller is set by controllerName');
+
+      assert.equal(text, 'home: myController',
+        'The homepage template was rendered with data from the custom controller');
+    });
+  }
+
+  [`@test The Homepage with a 'setupController' hook modifying other controllers`](assert) {
+    this.router.map(function() {
+      this.route('home', { path: '/' });
+    });
+
+    this.add('route:home', Route.extend({
+      setupController(/* controller */) {
+        set(this.controllerFor('home'), 'hours', emberA([
```

Comment:
Please just use

```js
this.controllerFor('home').set('hours', // ...
```

The `Ember.set` thing is just weird.

Reply:
For that matter, the use of `emberA([` instead of simply a normal array is unnecessary.

Reply:
thanks! (and I agree)